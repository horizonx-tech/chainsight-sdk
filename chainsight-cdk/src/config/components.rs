use std::collections::HashMap;

use anyhow::bail;
use lazy_static::lazy_static;
use regex::Regex;

use crate::indexer::IndexingConfig;

lazy_static! {
    static ref MAPPING_CANDID_TY: HashMap<&'static str, &'static str> = [
        ("text", "String"),
        // ("blob", "&[u8]"),
        ("nat", "u128"),
        ("int", "i128"),
        ("nat8", "u8"),
        ("nat16", "u16"),
        ("nat32", "u32"),
        ("nat64", "u64"),
        ("int8", "i8"),
        ("int16", "i16"),
        ("int32", "i32"),
        ("int64", "i64"),
        ("float32", "f32"),
        ("float64", "f64"),
        ("bool", "bool"),
        // ("null", "()"),
    ].iter().cloned().collect();

    static ref REGEX_CANDID_FUNC: Regex = Regex::new(r"(?P<identifier>\w+)\s*:\s*\((?P<params>.*?)\)\s*(->\s*\((?P<return>.*?)\))?").unwrap();
    static ref REGEX_VECTOR: Regex = Regex::new(r"vec\s(?P<type>\w+)").unwrap();
    static ref REGEX_TUPLE: Regex = Regex::new(r"record\s\{\s(?P<items>(\w+(;\s|))+)\s\}").unwrap();
    static ref REGEX_STRUCT: Regex = Regex::new(r"(?P<field>\w+)\s*:\s*(?P<type>\w+)").unwrap();

    static ref REGEX_MULTIPLE_RECORD_TYPE: Regex = Regex::new(r"record\s*\{").unwrap();
}

#[derive(serde::Deserialize, serde::Serialize)]
pub struct CommonConfig {
    pub monitor_duration: u32,
    pub canister_name: String,
}

#[derive(serde::Deserialize, serde::Serialize)]
pub struct AlgorithmIndexerInput {
    pub method_name: String,
    pub response_type: String,
    pub source_type: AlgorithmInputType,
}

#[derive(Clone, Debug, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum AlgorithmInputType {
    #[serde(rename = "event_indexer")]
    EventIndexer,
    #[serde(rename = "key_value")]
    KeyValue,
    #[serde(rename = "key_values")]
    KeyValues,
}

impl Default for AlgorithmIndexerInput {
    fn default() -> Self {
        Self {
            method_name: "get_list".to_string(),
            response_type: "String".to_string(),
            source_type: AlgorithmInputType::EventIndexer,
        }
    }
}

impl Default for CommonConfig {
    fn default() -> Self {
        Self {
            monitor_duration: 3600,
            canister_name: "example_canister".to_string(),
        }
    }
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct AlgorithmIndexerConfig {
    pub common: CommonConfig,
    pub indexing: IndexingConfig,
    pub input: AlgorithmIndexerInput,
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct AlgorithmLensConfig {
    pub common: CommonConfig,
    pub target_count: usize,
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct EventIndexerConfig {
    pub common: CommonConfig,
    pub def: EventIndexerEventDefinition,
}
#[derive(Default, Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct EventIndexerEventDefinition {
    pub identifier: String,
    pub abi_file_path: String,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct RelayerConfig {
    pub common: CommonConfig,
    pub destination: String,
    pub oracle_type: String,
    pub method_name: String,
    pub canister_method_value_type: CanisterMethodValueType,
    pub abi_file_path: String,
    pub lens_targets: Option<LensTargets>,
}
#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct LensTargets {
    pub identifiers: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub enum CanisterMethodValueType {
    Scalar(String, bool),                // struct name, is_scalar
    Tuple(Vec<(String, bool)>),          // struct_name, is_scalar
    Struct(Vec<(String, String, bool)>), // temp: Only non-nested `record` are supported.
    Vector(String, bool),                // struct_name, is_scalar
}
#[derive(Debug, Clone, PartialEq)]
pub struct CanisterMethodIdentifier {
    pub identifier: String,
    pub params: Vec<String>,
    pub return_value: CanisterMethodValueType,
}

impl CanisterMethodIdentifier {
    pub fn parse_from_str(s: &str) -> anyhow::Result<Self> {
        let captures = REGEX_CANDID_FUNC.captures(s).ok_or(anyhow::anyhow!(
            "method.identifier does not satisfy the supported expression: {}, supplied={}",
            REGEX_CANDID_FUNC.to_string(),
            s
        ))?;

        let identifier = captures.name("identifier").unwrap().as_str().to_string();

        let params_str = captures.name("params").unwrap().as_str();
        let params_result: Vec<String> = if params_str.is_empty() {
            vec![]
        } else {
            params_str
                .split(',')
                .map(|s| convert_type_from_candid_type(s.trim()).0)
                .collect()
        };
        let params = params_result;

        let return_value_str = captures.name("return").unwrap().as_str();
        let return_value = Self::parse_return_value(return_value_str)?;

        Ok(CanisterMethodIdentifier {
            identifier,
            params,
            return_value,
        })
    }

    fn parse_return_value(s: &str) -> anyhow::Result<CanisterMethodValueType> {
        let record_type_count = REGEX_MULTIPLE_RECORD_TYPE.find_iter(s).count();
        if record_type_count >= 2 {
            bail!(
                "Sorry, Currently nested `record` types are not supported. This will be supported in the future.\nTarget literal = {}",
                s
            ); // TODO: Support nested `record` types.
        }
        // vector
        if s.starts_with("vec") {
            let captures = REGEX_VECTOR.captures(s);
            if let Some(captures_value) = captures {
                let ty = captures_value.name("type").unwrap().as_str();
                let val = convert_type_from_candid_type(ty);
                return Ok(CanisterMethodValueType::Vector(val.0, val.1));
            }
            bail!("Invalid candid's result types:{}", s);
        }

        // Scalar
        if !s.starts_with("record") {
            let val = convert_type_from_candid_type(s);
            return Ok(CanisterMethodValueType::Scalar(val.0, val.1));
        }

        // Tuple
        let captures = REGEX_TUPLE.captures(s);
        if let Some(captures_value) = captures {
            let items = captures_value.name("items").unwrap().as_str();
            let tuple_result: Vec<(String, bool)> = items
                .split(';')
                .map(|s| convert_type_from_candid_type(s.trim()))
                .collect();
            let tuple = tuple_result;
            return Ok(CanisterMethodValueType::Tuple(tuple));
        }

        // Struct
        let items = REGEX_STRUCT.captures_iter(s);
        let mut struct_items = vec![];
        for cap in items {
            let field = cap.name("field").unwrap().as_str().to_string();
            let ty = convert_type_from_candid_type(cap.name("type").unwrap().as_str());
            struct_items.push((field, ty.0, ty.1));
        }
        if struct_items.is_empty() {
            bail!("Invalid candid's result types: {}", s);
        }
        Ok(CanisterMethodValueType::Struct(struct_items))
    }
}

fn convert_type_from_candid_type(s: &str) -> (String, bool) {
    // ref: https://internetcomputer.org/docs/current/references/candid-ref
    let ty_str = MAPPING_CANDID_TY.get(&s);
    if let Some(ty_str) = ty_str {
        return (ty_str.to_string(), true);
    }
    (s.to_string(), false)
}
