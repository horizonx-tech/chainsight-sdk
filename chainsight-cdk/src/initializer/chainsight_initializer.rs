use async_trait::async_trait;
use candid::{CandidType, Principal};
use ic_cdk::api::call::CallResult;

use super::{CycleManagements, InitConfig, InitError, InitResult, Initializer};

pub struct ChainsightInitializer {
    config: InitConfig,
}

impl ChainsightInitializer {
    pub fn new(config: InitConfig) -> Self {
        Self { config }
    }
}

#[derive(CandidType, serde::Deserialize, Clone, Copy)]
pub struct InitializeOutput {
    pub proxy: Principal,
    pub db: Principal,
}

#[async_trait]
impl Initializer for ChainsightInitializer {
    async fn initialize(&self, cycles: CycleManagements) -> Result<InitResult, InitError> {
        let out: CallResult<(InitializeOutput,)> = ic_cdk::api::call::call(
            self.config.env.initializer(),
            "initialize",
            (ic_cdk::caller(), cycles),
        )
        .await;
        Ok(InitResult {
            proxy: out.unwrap().0.proxy,
        })
    }
}
