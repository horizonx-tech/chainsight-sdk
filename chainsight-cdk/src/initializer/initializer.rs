use core::fmt;

use async_trait::async_trait;
use candid::{CandidType, Deserialize, Principal};

use crate::core::Env;
#[derive(Debug, CandidType)]
pub enum InitError {
    InvalidPrincipal(Principal),
    InvalidRequest(String),
    InvalidContent(String),
    InvalidDestination(String),
}
impl fmt::Display for InitError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            InitError::InvalidPrincipal(p) => write!(f, "Invalid principal: {}", p),
            InitError::InvalidRequest(s) => write!(f, "Invalid request: {}", s),
            InitError::InvalidContent(s) => write!(f, "Invalid content: {}", s),
            InitError::InvalidDestination(s) => write!(f, "Invalid destination: {}", s),
        }
    }
}
impl std::error::Error for InitError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            InitError::InvalidPrincipal(_) => None,
            InitError::InvalidRequest(_) => None,
            InitError::InvalidContent(_) => None,
            InitError::InvalidDestination(_) => None,
        }
    }
}

#[derive(CandidType, serde::Serialize, Deserialize, Clone, Copy, PartialEq, Debug)]
pub struct CycleManagement {
    pub initial_value: u128,
    pub refueling_value: u128,
    pub refueling_threshold: u128,
}

#[derive(CandidType, serde::Serialize, Deserialize, Clone, Copy, PartialEq, Debug)]
pub struct CycleManagements {
    pub refueling_interval: u64,
    pub vault_intial_supply: u128,
    pub indexer: CycleManagement,
    pub db: CycleManagement,
    pub proxy: CycleManagement,
}

impl CycleManagements {
    pub fn initial_supply(&self) -> u128 {
        self.vault_intial_supply
            + self.indexer.initial_value
            + self.db.initial_value
            + self.proxy.initial_value
    }
}

impl Default for CycleManagements {
    fn default() -> Self {
        Self {
            refueling_interval: 86400,
            vault_intial_supply: 1_000_000_000_000,
            indexer: CycleManagement {
                initial_value: 0,
                refueling_value: 1_000_000_000_000,
                refueling_threshold: 500_000_000_000,
            },
            db: CycleManagement {
                initial_value: 2_000_000_000_000,
                refueling_value: 1_00_000_000_000,
                refueling_threshold: 1_000_000_000_000,
            },
            proxy: CycleManagement {
                initial_value: 2_000_000_000,
                refueling_value: 1_000_000_000,
                refueling_threshold: 1_000_000_000,
            },
        }
    }
}

#[async_trait]
pub trait Initializer {
    async fn initialize(
        &self,
        deployer: &Principal,
        cycles: &CycleManagements,
    ) -> Result<InitResult, InitError>;
}
pub struct InitConfig {
    pub env: Env,
}

pub struct InitResult {
    pub proxy: Principal,
    pub db: Principal,
    pub vault: Principal,
}
