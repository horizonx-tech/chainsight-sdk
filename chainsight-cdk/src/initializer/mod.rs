mod initializer;
pub use initializer::*;
mod chainsight_initializer;
pub use chainsight_initializer::*;
