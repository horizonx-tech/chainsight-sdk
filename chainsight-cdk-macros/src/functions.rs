use chainsight_cdk::config::components::{LENS_FUNCTION_ARGS_TYPE, LENS_FUNCTION_RESPONSE_TYPE};
use proc_macro::TokenStream;
use quote::{format_ident, quote};
use syn::parse::{Parse, ParseStream};
use syn::LitInt;
use syn::{braced, punctuated::Punctuated, Ident, Result, Token, Type};

struct InitInEnvArgs {
    stable_memory_id: Option<LitInt>,
}
impl Parse for InitInEnvArgs {
    fn parse(input: ParseStream) -> Result<Self> {
        let stable_memory_id = if !input.is_empty() {
            let parsed: LitInt = input.parse()?;
            Some(parsed)
        } else {
            None
        };
        Ok(InitInEnvArgs { stable_memory_id })
    }
}
pub fn init_in_env(input: TokenStream) -> TokenStream {
    let args = syn::parse_macro_input!(input as InitInEnvArgs);
    init_in_env_internal(args).into()
}
fn init_in_env_internal(input: InitInEnvArgs) -> proc_macro2::TokenStream {
    let struct_quote = quote! {
        #[derive(Debug, Clone, Default, PartialEq, candid::CandidType, candid::Deserialize, serde::Serialize, StableMemoryStorable)]
        #[stable_mem_storable_opts(max_size = 10000, is_fixed_size = false)] // temp: max_size
        pub struct InitializingState {
            pub initialized: bool,
            pub proxy: String,
            pub env: chainsight_cdk::core::Env,
        }
    };
    let storage_quote = if let Some(memory_id) = input.stable_memory_id {
        quote! {
            stable_memory_for_scalar!("initializing_state", InitializingState, #memory_id, false);
        }
    } else {
        quote! {
            manage_single_state!("initializing_state", InitializingState, false);
        }
    };

    quote! {
        #struct_quote
        #storage_quote

        use chainsight_cdk::initializer::{CycleManagements, Initializer};
        use ic_cdk::api::management_canister::{provisional::{CanisterIdRecord, CanisterSettings}, main::{update_settings, UpdateSettingsArgument}};
        #[ic_cdk::update]
        #[candid::candid_method(update)]
        async fn init_in(env: chainsight_cdk::core::Env, cycles: CycleManagements) -> Result<(), chainsight_cdk::initializer::InitError> {
            assert!(!is_initialized(), "Already initialized");

            let initializer = chainsight_cdk::initializer::ChainsightInitializer::new(
                chainsight_cdk::initializer::InitConfig { env: env.clone() },
            );
            let deployer = ic_cdk::caller();
            let init_result = initializer.initialize(&deployer, &cycles).await?;
            let proxy = init_result.proxy;

            set_initializing_state(InitializingState {
                initialized: true,
                proxy: proxy.to_text(),
                env,
            }); // note: when using stable_memory, the returned result is not handled

            Ok(())
        }
        fn proxy() -> candid::Principal {
            candid::Principal::from_text(get_initializing_state().proxy).unwrap()
        }
        fn get_env() -> chainsight_cdk::core::Env {
            get_initializing_state().env
        }
        fn is_initialized() -> bool {
            get_initializing_state().initialized
        }

        #[ic_cdk::update]
        #[candid::candid_method(update)]
        fn get_proxy() -> candid::Principal {
            proxy()
        }
    }
}

struct SetupArgs {
    fields: Punctuated<NamedField, Token![,]>,
    stable_memory_id: Option<LitInt>,
}
impl Parse for SetupArgs {
    fn parse(input: ParseStream) -> Result<Self> {
        let content;
        braced!(content in input);
        let fields = Punctuated::parse_terminated(&content)?;
        let stable_memory_id = if input.peek(syn::Token![,]) {
            input.parse::<syn::Token![,]>()?;
            let parsed: LitInt = input.parse()?;
            Some(parsed)
        } else {
            None
        };

        Ok(SetupArgs {
            fields,
            stable_memory_id,
        })
    }
}
struct NamedField {
    name: Ident,
    _colon_token: Token![:],
    ty: Type,
}
impl Parse for NamedField {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(NamedField {
            name: input.parse()?,
            _colon_token: input.parse()?,
            ty: input.parse()?,
        })
    }
}
pub fn setup_func(input: TokenStream) -> TokenStream {
    let args = syn::parse_macro_input!(input as SetupArgs);
    setup_func_internal(args).into()
}
fn setup_func_internal(input: SetupArgs) -> proc_macro2::TokenStream {
    let SetupArgs {
        fields,
        stable_memory_id,
    } = input;

    let setters: Vec<_> = fields
        .iter()
        .map(|field| Ident::new(&format!("set_{}", field.name), field.name.span()))
        .collect();

    let names: Vec<_> = fields.iter().map(|field| &field.name).collect();
    let types: Vec<_> = fields.iter().map(|field| &field.ty).collect();

    // If stable_memory_id is specified, use Stable Memory, otherwise use Heap Memory
    let storage_quote = if let Some(memory_id) = stable_memory_id {
        quote! {
            stable_memory_for_scalar!("setup_flag", chainsight_cdk::storage::StorableBool, #memory_id, false);
        }
    } else {
        quote! {
            manage_single_state!("setup_flag", bool, false);
        }
    };

    quote! {
        #storage_quote

        #[ic_cdk::update]
        #[candid::candid_method(update)]
        fn setup(#( #names: #types ),*) -> Result<(), String> {
            assert!(!bool::from(get_setup_flag()), "Already setup");
            #( #setters(#names.into()); )*
            set_setup_flag(true.into()); // note: when using stable_memory, the returned result is not handled
            Ok(())
        }
    }
}

struct TimerTaskArgs {
    func_name: syn::LitStr,
    called_func_name: syn::LitStr,
    stable_memory_id: Option<LitInt>,
}
impl Parse for TimerTaskArgs {
    fn parse(input: ParseStream) -> Result<Self> {
        let func_name: syn::LitStr = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let called_func_name: syn::LitStr = input.parse()?;
        let stable_memory_id = if input.peek(syn::Token![,]) {
            input.parse::<syn::Token![,]>()?;
            let parsed: LitInt = input.parse()?;
            Some(parsed)
        } else {
            None
        };
        Ok(TimerTaskArgs {
            func_name,
            called_func_name,
            stable_memory_id,
        })
    }
}
pub fn timer_task_func(input: TokenStream) -> TokenStream {
    let args = syn::parse_macro_input!(input as TimerTaskArgs);
    timer_task_func_internal(args).into()
}
fn timer_task_func_internal(args: TimerTaskArgs) -> proc_macro2::TokenStream {
    let TimerTaskArgs {
        func_name,
        called_func_name,
        stable_memory_id,
    } = args;

    let func_name_ident = format_ident!("{}", func_name.value());
    let called_func_name_str = called_func_name.value().to_string();

    // If stable_memory_id is specified, use Stable Memory, otherwise use Heap Memory
    let storage_quote = if let Some(memory_id) = stable_memory_id {
        quote! {
            stable_memory_for_scalar!("indexing_interval", u32, #memory_id, false);
        }
    } else {
        quote! {
            manage_single_state!("indexing_interval", u32, false);
        }
    };

    quote! {
        #storage_quote

        #[ic_cdk::update]
        #[candid::candid_method(update)]
        pub async fn #func_name_ident(task_interval_secs: u32, delay_secs: u32, is_rounded_start_time: bool) {
            set_indexing_interval(task_interval_secs);
            let res = ic_cdk::api::call::call::<(u32, u32, bool, String, Vec<u8>), ()>(
                proxy(),
                "start_indexing_with_is_rounded",
                (task_interval_secs, delay_secs, is_rounded_start_time, #called_func_name_str.to_string(), Vec::<u8>::new()),
            )
            .await;
            match res {
                Ok(_) => {}
                Err(e) => { panic!("Failed to start indexing: {:?}", e) }
            };
        }
    }
}

#[derive(Debug)]
struct LensMethodArgs {
    target_count: usize,
    func_arg: Option<Type>,
}
impl Parse for LensMethodArgs {
    fn parse(input: ParseStream) -> Result<Self> {
        let target_count: syn::LitInt = input.parse()?;
        let func_arg = if input.peek(syn::Token![,]) {
            input.parse::<syn::Token![,]>()?;
            Some(input.parse()?)
        } else {
            None
        };
        Ok(LensMethodArgs {
            target_count: target_count.base10_parse().unwrap(),
            func_arg,
        })
    }
}
pub fn lens_method(input: TokenStream) -> TokenStream {
    let args = syn::parse_macro_input!(input as LensMethodArgs);
    lens_method_internal(args).into()
}
fn lens_method_internal(args: LensMethodArgs) -> proc_macro2::TokenStream {
    let getter_name = format_ident!("{}", "get_result");
    let proxy_getter_name = format_ident!("{}", "proxy_get_result");

    let args_ty = format_ident!("{}", LENS_FUNCTION_ARGS_TYPE);
    let value_ty = format_ident!("{}", LENS_FUNCTION_RESPONSE_TYPE);
    let target_count = args.target_count;

    let (lens_args_def, getter_func, receiver_provider, inter_calc_func) = match args.func_arg {
        Some(func_arg_ty) => (
            quote! {
                #[derive(Clone, Debug, Default, candid::CandidType, serde::Deserialize, serde::Serialize)]
                pub struct #args_ty {
                    pub targets: Vec<String>,
                    pub args: #func_arg_ty,
                }
            },
            quote! {
                #[ic_cdk::update]
                #[candid::candid_method(update)]
                async fn #getter_name(input: #args_ty) -> #value_ty {
                    if input.targets.len() != #target_count {
                        panic!("Expected {} targets", #target_count);
                    }
                    _calc(input).await
                }
            },
            quote! {
                chainsight_cdk::rpc::AsyncReceiverProvider::<#args_ty, #value_ty>::new(
                    proxy(),
                    _calc,
                )
            },
            quote! {
                fn _calc(input: #args_ty) -> BoxFuture<'static, #value_ty> {
                    async move { calculate(input.targets, input.args).await }.boxed()
                }
            },
        ),
        None => (
            quote! {
                type #args_ty = Vec<String>;
            },
            quote! {
                #[ic_cdk::update]
                #[candid::candid_method(update)]
                async fn #getter_name(targets: #args_ty) -> #value_ty {
                    if targets.len() != #target_count {
                        panic!("Expected {} targets", #target_count);
                    }
                    _calc(targets).await
                }
            },
            quote! {
                chainsight_cdk::rpc::AsyncReceiverProvider::<#args_ty, #value_ty>::new(
                    proxy(),
                    _calc,
                )
            },
            quote! {
                fn _calc(input: #args_ty) -> BoxFuture<'static, #value_ty> {
                    async move { calculate(input).await }.boxed()
                }
            },
        ),
    };

    quote! {
        #lens_args_def
        #getter_func

        #[ic_cdk::update]
        #[candid::candid_method(update)]
        async fn #proxy_getter_name(input:Vec<u8>) -> Vec<u8> {
            use chainsight_cdk::rpc::Receiver;
            let reciever_provider = #receiver_provider;
            reciever_provider.reply(input).await
        }

        #[ic_cdk::query]
        #[candid::candid_method(query)]
        fn get_sources() -> Vec<chainsight_cdk::core::Sources<std::collections::HashMap<String, String>>> {
            vec![]
        }

        #inter_calc_func
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;
    use rust_format::{Formatter, RustFmt};

    use super::*;

    #[test]
    fn test_snapshot_init_in_env() {
        let input = quote! {};
        let args: syn::Result<InitInEnvArgs> = syn::parse2(input);
        let generated = init_in_env_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__init_in_env", formatted);
    }

    #[test]
    fn test_snapshot_init_in_env_with_stable_memory() {
        let input = quote! {1};
        let args: syn::Result<InitInEnvArgs> = syn::parse2(input);
        let generated = init_in_env_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__init_in_env_with_stable_memory", formatted);
    }

    #[test]
    fn test_snapshot_setup_func() {
        let input = quote! {
            {
                target_canister: String,
                target_addr: String,
                web3_ctx_param: Web3CtxParam
            }
        };
        let args: syn::Result<SetupArgs> = syn::parse2(input);
        let generated = setup_func_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__setup_func", formatted);
    }

    #[test]
    fn test_snapshot_setup_func_with_stable_memory() {
        let input = quote! {
            {
                target_canister: String,
                target_addr: String,
                web3_ctx_param: Web3CtxParam
            },
            0
        };
        let args: syn::Result<SetupArgs> = syn::parse2(input);
        let generated = setup_func_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__setup_func_with_stable_memory", formatted);
    }

    #[test]
    fn test_snapshot_timer_task_func() {
        let input = quote! {"set_task","HELLO"};
        let args: syn::Result<TimerTaskArgs> = syn::parse2(input);
        let generated = timer_task_func_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__timer_task_func", formatted);
    }

    #[test]
    fn test_snapshot_timer_task_func_with_stable_memory() {
        let input = quote! {"set_task","HELLO",1};
        let args: syn::Result<TimerTaskArgs> = syn::parse2(input);
        let generated = timer_task_func_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__timer_task_func_with_stable_memory", formatted);
    }

    #[test]
    fn test_snapshot_lens_method() {
        let input = quote! {10};
        let args: syn::Result<LensMethodArgs> = syn::parse2(input);
        let generated = lens_method_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__lens_method", formatted);
    }

    #[test]
    fn test_snapshot_lens_method_with_args() {
        let input = quote! {10, CalculateArgs};
        let args: syn::Result<LensMethodArgs> = syn::parse2(input);
        let generated = lens_method_internal(args.unwrap());
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__lens_method__with_args", formatted);
    }
}
