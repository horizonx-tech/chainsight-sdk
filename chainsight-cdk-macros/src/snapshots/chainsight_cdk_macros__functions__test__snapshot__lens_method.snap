---
source: chainsight-cdk-macros/src/functions.rs
expression: formatted
---
#[ic_cdk::update]
#[candid::candid_method(update)]
async fn get_result(targets: Vec<String>) -> LensValue {
    if targets.len() != 10usize {
        panic!("Expected {} targets", 10usize);
    }
    _calc(targets).await
}
#[ic_cdk::update]
#[candid::candid_method(update)]
async fn proxy_get_result(input: Vec<u8>) -> Vec<u8> {
    use chainsight_cdk::rpc::Receiver;
    let reciever_provider =
        chainsight_cdk::rpc::AsyncReceiverProvider::<Vec<String>, LensValue>::new(proxy(), _calc);
    reciever_provider.reply(input).await
}
#[ic_cdk::query]
#[candid::candid_method(query)]
fn get_sources() -> Vec<chainsight_cdk::core::Sources<std::collections::HashMap<String, String>>> {
    vec![]
}
fn _calc(targets: Vec<String>) -> BoxFuture<'static, LensValue> {
    async move { calculate(targets).await }.boxed()
}

